### ISSUE TITLE
FAP03

### ISSUE THREAD

**Issue by Marti-creator** [Marti-creator/DeepLearning#1]

**Comment by Marti-creator** [comment]
# -*- coding: utf-8 -*-
"""2324_FAP_Lab_03_IMDB.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/11Pofr9M2vo2XpuB902ZMah5ZtLVLGwO2

## Classifying movie reviews: A binary classification example
*# Credits to F. Chollet. Introduction to Deep Learning*
Edited by eduardo.fidalgo@unileon.es for academic purposes

Import the required libraries
"""

# Import the required libraries for the laboratory session.
import numpy as np
from keras.datasets import imdb
from keras import models
from keras import layers



"""Data loading: **IMDB** dataset

"""

# This time we are going to load the full dataset. Notice that in Lab01 we
# did not include the first argument of this function, we did something like
# _, (test_data, test_labels), to work only with the test data
(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words=10000)

print ("%%% Exercises %%%")
# Exercise 01: get some insights from IMDB dataset. Use the train/test variables
# and obtain (i) its type, (ii) tensor rank or dimension, (iii) shape.
# Exercise 01: solution printed
print ('Exercise 01. The type of train_data is ' +
       str(type(train_data)) + " with " +
       str(train_data.ndim) + " dimensions " +
       "and a shape of " + str(train_data.shape))

# Exercise 02: how many reviews does IMDB dataset have?
# Reply completing the python code to retrieve from images np.array the
# variable that contains this number.
print ('Exercise 02. IMDB dataset contains ' + str(len(train_data)) + " reviews")

"""Displaying how it looks a review of IMDB coded in BoW"""

# Exercise 03: visualize how  it looks the first element of the train_data tensor
first_element = train_data[0]
print ('Exercise 03. This is the first element of train_data ' + str(first_element))

"""**Decoding reviews back to text**"""

word_index = imdb.get_word_index()
reverse_word_index = dict(
    [(value, key) for (key, value) in word_index.items()])
decoded_review = " ".join(
    [reverse_word_index.get(i - 3, "?") for i in train_data[0]])

# Exercise 04: print the first 10 elements of the word_index dictionary
# Noticed how each ID corresponds to a word.
print ('Exercise 04.')
import operator

small_word_index = dict(sorted(word_index.items(), key=operator.itemgetter(1))[:10])

for words in small_word_index:
  word = small_word_index[words]
  print('The ID %s corresponds to the word %s' % (words, word))

"""### Preparing the data

**Encoding the data to fed the neural network**
"""

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        for j in sequence:
            results[i, j] = 1.
    return results
x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

# Exercise 05: print the first element of train_data and x_train.
# See their diferences. Explain
# why are we doing this conversion of the data?
print(train_data[0])
print(x_train[0])

print ('Exercise 05. We are encoding the integers sequences because',\
       'neural networks require numerical inputs.')


# Exercise 06: mention the two possible ways of preparing data for a NN
print ('Exercise 06. Two possible ways to prepare data for a neural network',\
 'are to vectorizing the data or using embedding layers.')

# Do not forget that labels must be encoded too!!
y_train = np.asarray(train_labels).astype("float32")
y_test = np.asarray(test_labels).astype("float32")

"""### Building your model

**Model definition**
"""

# Exercise 07. Build a keras model (Sequential) with three dense layers.
# Layer 1 -> 16 neurons and ReLU activation.
# Layer 2 -> 16 neurons and ReLU activation.
# Layer 3 -> 1 neuron and sigmoid activation.
print ('Exercise 07. ')
from keras.models import Sequential
from keras.layers import Dense
model = Sequential([
    Dense(16, activation='relu', input_shape=(10000,)),
    Dense(16, activation='relu'),
    Dense(1,activation='sigmoid')
])

print ('Model build!! ')

"""**Compiling the model**"""

# Exercise 08. Compile the model using an optimizer that implements
# RMSprop algorithm,
# Use a Cross Entropy loss function
# Select accuracy as the metric to be evaluated by the model during training/test
print ('Exercise 08. ')
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

"""### Validating your approach

**Setting aside a validation set**
"""

# Exercise 09. From the training set, separate the first 10000 reviews as a
#a validation set (x_val), together with their labels (y_val).
print ('Exercise 09. ')
x_val = x_train[:10000]
partial_x_train = x_train[10000:]
y_val = y_train[:10000]
partial_y_train = y_train[10000:]

"""**Training your model**"""

# Exercise 10. Train your model using the partial_x_train training set you splitted,
# together with its labels (partial_y_train).
# Train it for 20 epoch and select a batch size of
# 512. Use as validation data the one you created before.
print ('Exercise 10. ')
history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()

"""**Plotting the training and validation loss**"""

# Exercise 11. Fill the missing XXXX
print('Exercise 11.')
import matplotlib.pyplot as plt

# Take from the dictionary the values you are going to plot. The history is
# and object (dictionary) containing "everything" happened during training
history_dict = history.history
loss_values = history_dict["loss"]
val_loss_values = history_dict["val_loss"]

# Create a list containing the number of epochs, so you can see how the training
# evolves per epoch
epochs = range(1, len(loss_values) + 1)

# Plot customization, labeling and display (e.g. bo -> blue dot)
plt.plot(epochs, loss_values, "bo", label="Training loss")
plt.plot(epochs, val_loss_values, "b", label="Validation loss")
plt.title("Training and Validation loss")
plt.xlabel("Epochs")
plt.ylabel("Loss")
plt.legend()
plt.show()

"""**Plotting the training and validation accuracy**"""

plt.clf()
# Exercise 12. Plot the training (acc) and validation accuracy (val_acc)
acc = history_dict["accuracy"]
val_acc = history_dict["val_accuracy"]
plt.plot(epochs, acc, "bo", label="Training acc")
plt.plot(epochs, val_acc, "b", label="Validation acc")
plt.title("Training and validation accuracy")
plt.xlabel("Epochs")
plt.ylabel("Accuracy")
plt.legend()
plt.show()

"""**Obtaining the performance on the test set**"""

results = model.evaluate(x_test, y_test)
print("test loss, test acc:", results)

"""**Retraining a model from scratch and obtain newer results**"""

from tensorflow import keras

model = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
x = 10
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

model.fit(x_train, y_train, epochs=x, batch_size=512)

results = model.evaluate(x_test, y_test)
print("test loss, test acc:", results)
